### PR TITLE
Param MATCH vid seek fix

### PR DESCRIPTION
**Pull request by wey-gu**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### Description:
Fixed this case:

```cypher
:param p3 => "player100"
:param p8 => "player101"
:param p7 => {"a":{"b":{"c":"Tim Duncan","d":[1,2,3,true,"player100"]}}}

      MATCH (v) WHERE id(v) == $p3
      RETURN id(v) AS v

      MATCH (v) WHERE id(v) IN [$p3,$p8]
      RETURN id(v) AS v

      MATCH (v) WHERE id(v) == $p7.a.b.d[4]
      RETURN id(v) AS v

      MATCH (v) WHERE id(v) IN $p7.a.b.d
      RETURN id(v) AS v
```

## How do you solve it?

To make right part of the expression evaluated in `src/graph/visitor/VidExtractVisitor.cpp`

## Checklist:
Tests:
- [x] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [x] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

Fixed the bug on parameters for `id(n) == $var ` , `id(n) IN [$var]`, `id(n) == $var.foo.bar`, `id(n) IN $var.foo.bar`

### CODE REVIEW

**Review by CPWstatic — COMMENTED** [review]
Good job! Does it work now? I'm looking forward to the test cases.

**Comment by wey-gu** [comment]
> Good job! Does it work now? I'm looking forward to the test cases.

Both `WHERE id(n) == $var ` and `WHERE id(n) IN [$var]` works now, OK, will add test cases later.

**Comment by czpmango** [comment]
Please add tck test cases.

File: src/graph/visitor/VidExtractVisitor.cpp [hunk 1]
```diff
@@ -165,12 +165,10 @@ void VidExtractVisitor::visit(RelationalExpression *expr) {
       return;
     }
 
-    auto *listExpr = static_cast<ListExpression *>(expr->right());
```
Comment (czpmango) [inline src/graph/visitor/VidExtractVisitor.cpp:168]:
```
DCHECK(rightListValue.isList());
```
Some related situations need to be concerned.

**Comment by czpmango** [comment]
> > Good job! Does it work now? I'm looking forward to the test cases.
> 
> Both `WHERE id(n) == $var ` and `WHERE id(n) IN [$var]` works now, OK, will add unit test cases later in separated PR.

What is the behavior in this case?
```
match (n) where id(n) in [$var1,$var2...] return n
```

File: src/graph/visitor/VidExtractVisitor.cpp (hunk 1 above)
Comment (wey-gu) [reply src/graph/visitor/VidExtractVisitor.cpp:168]: > ```
> DCHECK(rightListValue.isList());
> ```
> 
> Some related situations need to be concerned.

Thanks, will add this check.

**Comment by wey-gu** [comment]
> > > Good job! Does it work now? I'm looking forward to the test cases.
> > 
> > 
> > Both `WHERE id(n) == $var ` and `WHERE id(n) IN [$var]` works now, OK, will add unit test cases later in separated PR.
> 
> What is the behavior in this case?
> 
> ```
> match (n) where id(n) in [$var1,$var2...] return n
> ```

Thanks for the reminding.

```
(root@nebula) [basketballplayer]> :param v1 => "player101"

Wed, 30 Mar 2022 15:12:24 UTC

(root@nebula) [basketballplayer]> return $v0, $v1
+-------------+-------------+
| $v0         | $v1         |
+-------------+-------------+
| "player100" | "player101" |
+-------------+-------------+
Got 1 rows (time spent 2113/3169 us)

Wed, 30 Mar 2022 15:12:26 UTC

(root@nebula) [basketballplayer]> match (n) where id(n) in [$v0, $v1] return n
+-----------------------------------------------------+
| n                                                   |
+-----------------------------------------------------+
| ("player100" :player{age: 42, name: "Tim Duncan"})  |
| ("player101" :player{age: 36, name: "Tony Parker"}) |
+-----------------------------------------------------+
Got 2 rows (time spent 10833/12107 us)

```

**Comment by wey-gu** [comment]
> Please add tck test cases.

I am preparing for the tck cases, while it's strange that this case passes.

https://github.com/vesoft-inc/nebula/blob/1e68162c6e46b601575cd708e296538b7db884d8/tests/tck/features/yield/parameter.feature#L19-L28

The behavior from console is shown as this:
```
(root@nebula) [basketballplayer]> :param var => "player100"
(root@nebula) [basketballplayer]> match (n) where id(n) == $var return n
[ERROR (-1005)]: Scan vertices or edges need to specify a limit number, or limit number can not push down.

Wed, 30 Mar 2022 23:28:08 CST
```

**Comment by wey-gu** [comment]
> Please add tck test cases.

done
```bash
root@1827b82e88bf:/home/nebula/tests# python3 -m pytest -m "wey"
===================================================================================== test session starts =====================================================================================
platform linux -- Python 3.8.10, pytest-5.3.2, py-1.11.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
benchmark: 3.2.3 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
metadata: {'Python': '3.8.10', 'Platform': 'Linux-5.4.0-94-generic-x86_64-with-glibc2.29', 'Packages': {'pytest': '5.3.2', 'py': '1.11.0', 'pluggy': '0.13.1'}, 'Plugins': {'reportlog': '0.1.0', 'benchmark': '3.2.3', 'forked': '1.4.0', 'metadata': '1.8.0', 'yapf3': '0.5.1', 'html': '2.0.1', 'xdist': '1.31.0', 'bdd': '4.0.2', 'drop-dup-tests': '0.3.0'}}
rootdir: /home/nebula/tests, inifile: pytest.ini
plugins: reportlog-0.1.0, benchmark-3.2.3, forked-1.4.0, metadata-1.8.0, yapf3-0.5.1, html-2.0.1, xdist-1.31.0, bdd-4.0.2, drop-dup-tests-0.3.0
collecting ... Generating LALR tables
collected 1354 items / 1350 deselected / 4 selected

tck/steps/test_tck.py::test_return_parameters PASSED                                                                                                                                    [ 25%]
tck/steps/test_tck.py::test_cypher_with_parameters PASSED                                                                                                                               [ 50%]
tck/steps/test_tck.py::test_ngql_with_parameters PASSED                                                                                                                                 [ 75%]
tck/steps/test_tck.py::test_error_check_2 PASSED                                                                                                                                        [100%]

======================================================================= 4 passed, 1350 deselected, 4 warnings in 1.27s ========================================================================
```

**Comment by wey-gu** [comment]
I see mem leak in [ci](https://github.com/vesoft-inc/nebula/runs/5757421449?check_suite_focus=true), it's not relevant to this PR, right?

**Review by CPWstatic — DISMISSED** [review]

File: tests/tck/features/yield/parameter.feature [hunk 2]
```diff
@@ -17,6 +17,23 @@ Feature: Parameter
...
+      | "Tim Duncan" |
+    When executing query:
+      """
+      MATCH (v) WHERE id(v) IN [$p3,$p8]
+      RETURN id(v) AS v
+      """
+    Then the result should be, in any order:
+      | v               |
+      | "Tim Duncan"    |
+      | "Manu Ginobili" |
```
Comment (czpmango) [inline tests/tck/features/yield/parameter.feature:36]: Add test cases:
```
MATCH (v) WHERE id(v) IN $p7.a.b.d RETURN v
```

**Comment by czpmango** [comment]
> I see mem leak in [ci](https://github.com/vesoft-inc/nebula/runs/5757421449?check_suite_focus=true), it's not relevant to this PR, right?

Yep.

File: tests/tck/features/yield/parameter.feature (hunk 2 above)
Comment (wey-gu) [reply tests/tck/features/yield/parameter.feature:36]: Thanks @czpmango , this case addressed the new case that was not yet supported and now it's fixed.

done.

**Comment by wey-gu** [comment]
It's now rebased with all comments resolved. I will trigger review again when full CI passed.

File: tests/tck/features/yield/parameter.feature (hunk 2 above)
Comment (czpmango) [reply tests/tck/features/yield/parameter.feature:36]: Nice.

**Review by czpmango — APPROVED** [review]
LGTM. Nice fix!

**Comment by wey-gu** [comment]
@CPWstatic , two more commits since last review

- https://github.com/vesoft-inc/nebula/pull/4024/commits/a3fb92383f3db532f018dd2b8dde5aa49d5c2a23 support kAttribute for in case
  - `MATCH (v) WHERE id(v) IN $p7.a.b.d RETURN v`
- https://github.com/vesoft-inc/nebula/pull/4024/commits/bf6161e1068164fa134ba80db15e6f3ce9aa03d4 Add kSubscript for eq case
  - `MATCH (v) WHERE id(v) == $p7.a.b.d[4] RETURN v`
  
@czpmango , one more commit since last review

- https://github.com/vesoft-inc/nebula/pull/4024/commits/bf6161e1068164fa134ba80db15e6f3ce9aa03d4 Add kSubscript for eq case
  - `MATCH (v) WHERE id(v) == $p7.a.b.d[4] RETURN v`

Thanks!

**Comment by wey-gu** [comment]
CI failed due to:

- CentOS gcc, 69 - log_command_test (Failed)
- Ubuntu gcc, exit code 137: OOM
- Ubuntu clang, mem leak, known issue

**Review by czpmango — APPROVED** [review]

**Review by CPWstatic — APPROVED** [review]

**Pull request merged**